consensus/src/rand/rand_gen/block_queue.rs

```diff
@@ -0,0 +1,248 @@
+            offsets_by_round,
+            num_undecided_blocks: len,
+        }
+    }
+
+    pub fn num_blocks(&self) -> usize {
+        self.blocks().len()
+    }
+
+    pub fn first_round(&self) -> u64 {
+        self.blocks().first().unwrap().block().round()
+    }
+
+    pub fn offset(&self, round: Round) -> usize {
+        *self.offsets_by_round.get(&round).unwrap()
+    }
+
+    pub fn num_undecided(&self) -> usize {
+        self.num_undecided_blocks
+    }
+
+    pub fn rand_metadata(&self, round: Round) -> RandMetadata {
+        let block = self.blocks()[self.offset(round)].block();
+        block.into()
+    }
+
+    pub fn set_randomness(&mut self, round: Round, rand: Randomness) -> bool {
+        let offset = self.offset(round);
+        if !self.blocks()[offset].has_randomness() {
+            self.blocks_mut()[offset].set_randomness(rand);
+            self.num_undecided_blocks -= 1;
+            true
+        } else {
+            false
+        }
+    }
+
+    fn blocks(&self) -> &[ExecutedBlock] {
+        &self.ordered_blocks.ordered_blocks
+    }
+
+    fn blocks_mut(&mut self) -> &mut [ExecutedBlock] {
+        &mut self.ordered_blocks.ordered_blocks
+    }
+}
+
+/// Maintain ordered blocks that have pending randomness
+pub struct BlockQueue {
+    queue: BTreeMap<Round, QueueItem>,
+}
+impl BlockQueue {
+    pub fn new() -> Self {
+        Self {
+            queue: BTreeMap::new(),
+        }
+    }
+
+    pub fn push_back(&mut self, item: QueueItem) {
+        assert!(self.queue.insert(item.first_round(), item).is_none());
+    }
+
+    /// Dequeue all ordered blocks prefix that have randomness
+    pub fn dequeue_rand_ready_prefix(&mut self) -> Vec<OrderedBlocks> {
+        let mut rand_ready_prefix = vec![];
+        while let Some((_starting_round, item)) = self.queue.first_key_value() {
+            if item.num_undecided() == 0 {
+                let (_, item) = self.queue.pop_first().unwrap();
+                let QueueItem { ordered_blocks, .. } = item;
+                debug_assert!(ordered_blocks
+                    .ordered_blocks
+                    .iter()
+                    .all(|block| block.has_randomness()));
+                rand_ready_prefix.push(ordered_blocks);
+            } else {
+                break;
+            }
+        }
+        rand_ready_prefix
+    }
+
+    /// Return the `QueueItem` that contains the given round, if exists.
+    pub fn item_mut(&mut self, round: Round) -> Option<&mut QueueItem> {
+        self.queue
+            .range_mut(0..=round)
+            .last()
+            .map(|(_, item)| item)
+            .filter(|item| item.offsets_by_round.contains_key(&round))
+    }
+
+    /// Update the corresponding block's randomness, return true if updated successfully
+    pub fn set_randomness(&mut self, round: Round, randomness: Randomness) -> bool {
+        if let Some(item) = self.item_mut(round) {
+            item.set_randomness(round, randomness)
```

Comment:
Do we expect this to fail? Are you going to use the return value somewhere?

Reply:
this is not expected to fail, but we may want to use the return value for logging purpose (if it's the first time or not)